File: python/tests/test_tools.py
"""Tests for src/tools.py — Week 4"""
import json
import pytest
from src.tools import (
    get_build_status, get_recent_deploys, get_active_incidents,
    execute_tool_safely, run_tool_loop, run_tool_loop_with_trace,
    ALL_TOOLS, TOOLS_BY_NAME, build_status, flaky,
)


# ═══════════════════════════════════════════════════════════════
# Tool definitions
# ═══════════════════════════════════════════════════════════════

def test_tools_imports_llm():
    """tools.py imports from llm.py — the import graph holds."""
    from src.tools import run_tool_loop
    import inspect
    source = inspect.getsource(run_tool_loop)
    assert "get_llm" in source, "run_tool_loop should call get_llm() from src.llm"


def test_get_build_status_known_service():
    """Returns status for a known service."""
    result = json.loads(get_build_status.invoke({"service_name": "auth-service"}))
    assert result["status"] == "healthy"
    assert "last_deploy" in result


def test_get_build_status_unknown_service():
    """Returns unknown for an unrecognized service."""
    result = json.loads(get_build_status.invoke({"service_name": "nonexistent"}))
    assert result["status"] == "unknown"
    assert "error" in result


def test_get_recent_deploys_returns_list():
    """Returns a list of deployment records."""
    result = json.loads(get_recent_deploys.invoke({
        "service_name": "payment-api", "limit": 3
    }))
    assert isinstance(result, list)
    assert len(result) == 3
    assert result[0]["status"] == "success"


def test_get_recent_deploys_empty_for_unknown():
    """Returns empty list for service with no deploys."""
    result = json.loads(get_recent_deploys.invoke({
        "service_name": "inventory-service"
    }))
    assert result == []


def test_get_active_incidents_has_incidents():
    """Returns active incidents for payment-api."""
    result = json.loads(get_active_incidents.invoke({
        "service_name": "payment-api"
    }))
    assert len(result) == 1
    assert result[0]["id"] == "INC-842"


def test_get_active_incidents_none_for_healthy():
    """Returns empty list for service with no incidents."""
    result = json.loads(get_active_incidents.invoke({
        "service_name": "auth-service"
    }))
    assert result == []


# ═══════════════════════════════════════════════════════════════
# Tool execution
# ═══════════════════════════════════════════════════════════════

def test_execute_tool_safely_known_tool():
    """Executes a valid tool call successfully."""
    result = json.loads(execute_tool_safely({
        "name": "get_build_status",
        "args": {"service_name": "auth-service"},
    }))
    assert result["status"] == "healthy"


def test_execute_tool_safely_unknown_tool():
    """Returns structured error for unknown tool."""
    result = json.loads(execute_tool_safely({
        "name": "nonexistent_tool",
        "args": {},
    }))
    assert "error" in result
    assert "available_tools" in result


def test_flaky_fails_first_n_then_succeeds():
    """flaky() is a deterministic, per-instance failure injector."""
    fn = flaky(build_status, fail_first_n=2)
    for _ in range(2):
        with pytest.raises(ConnectionError):
            fn("auth-service")
    assert json.loads(fn("auth-service"))["status"] == "healthy"


def test_flaky_instances_are_isolated():
    """Each flaky() wrapper owns its own counter — no shared global state."""
    a = flaky(build_status, fail_first_n=1)
    b = flaky(build_status, fail_first_n=1)
    with pytest.raises(ConnectionError):
        a("auth-service")
    # b's counter is untouched by a's failure
    with pytest.raises(ConnectionError):
        b("auth-service")
    assert json.loads(a("auth-service"))["status"] == "healthy"


def test_execute_tool_safely_denies_unknown_via_registry(monkeypatch):
    """The whitelist guard: an unregistered tool name is denied, not executed."""
    def _noop(**kwargs):
        raise AssertionError("should never be called")
    registry = {"get_build_status": object()}  # not even a real tool
    monkeypatch.setattr("time.sleep", lambda s: None)
    result = json.loads(execute_tool_safely(
        {"name": "delete_production_db", "args": {}},
        tools_by_name=registry,
    ))
    assert result["error"] == "Unknown tool: 'delete_production_db'"
    assert result["available_tools"] == ["get_build_status"]


def test_execute_tool_safely_exhausted_returns_structured_error(monkeypatch):
    """Persistent failure → structured error after retries burn out."""
    from langchain_core.tools import tool as lc_tool

    impl = flaky(build_status, fail_first_n=99)

    @lc_tool
    def get_build_status(service_name: str) -> str:
        """Return build status."""
        return impl(service_name)

    monkeypatch.setattr("time.sleep", lambda s: None)
    result = json.loads(execute_tool_safely(
        {"name": "get_build_status", "args": {"service_name": "payment-api"}},
        max_retries=1,
        tools_by_name={"get_build_status": get_build_status},
    ))
    assert result["status"] == "failed"
    assert result["attempts"] == 2  # 1 initial + 1 retry
    assert "error" in result


def test_all_tools_have_descriptions():
    """Every tool has a non-empty docstring for routing."""
    for t in ALL_TOOLS:
        assert t.description, f"Tool '{t.name}' has no description"


def test_tools_by_name_maps_all():
    """TOOLS_BY_NAME contains all tools."""
    assert set(TOOLS_BY_NAME.keys()) == {t.name for t in ALL_TOOLS}


# ═══════════════════════════════════════════════════════════════
# Tool-calling loop (requires LLM)
# ═══════════════════════════════════════════════════════════════

def test_run_tool_loop_calls_tool():
    """A question that needs build status triggers a tool call, and the answer
    is grounded in the tool result (in the model's own words, not verbatim)."""
    trace = run_tool_loop_with_trace("Is the auth-service healthy?", temperature=0.0)
    calls = [tc["name"] for tc in trace.get("tool_calls", [])]
    assert "get_build_status" in calls, f"Expected get_build_status call, got: {calls}"
    assert trace["tool_calls"][0]["args"].get("service_name") == "auth-service", (
        f"Tool should target auth-service: {trace['tool_calls'][0]['args']}"
    )
    answer = trace["answer"]
    assert len(answer) > 10, "Answer is too short"
    # The answer must reflect the tool result: a status word, the service name,
    # or the deploy marker that the tool returned. Brittle word-for-word
    # matching ("healthy") fails on correct-but-paraphrased answers.
    low = answer.lower()
    assert any(w in low for w in ["auth", "healthy", "degraded", "down", "unknown", "08:15"]), (
        f"Answer should reflect the build-status result: {answer[:120]}"
    )


def test_run_tool_loop_no_tool_needed():
    """A simple question doesn't trigger any tool calls."""
    result = run_tool_loop("What is 2 + 2?", temperature=0.0)
    assert "4" in result


def test_run_tool_loop_with_trace_returns_trace():
    """The traced version returns a dict with steps."""
    trace = run_tool_loop_with_trace(
        "Is the payment-api healthy?", temperature=0.0
    )
    assert "answer" in trace
    assert "query" in trace
    assert "steps" in trace
    assert len(trace["steps"]) >= 1


def test_run_tool_loop_with_trace_no_tool_calls():
    """Trace works even when no tools are called."""
    trace = run_tool_loop_with_trace("Hello!", temperature=0.0)
    assert "answer" in trace
    assert trace.get("tool_calls") is None or trace["tool_calls"] == []

File: nodejs/tests/test_tools.js
/**
 * Tests for src/tools.js — Week 4 (Node.js)
 */
import { describe, it, expect } from "vitest";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import {
  getBuildStatus,
  getRecentDeploys,
  getActiveIncidents,
  buildStatus,
  recentDeploys,
  activeIncidents,
  flaky,
  ConnectionError,
  executeToolSafely,
  runToolLoop,
  runToolLoopWithTrace,
  ALL_TOOLS,
  TOOLS_BY_NAME,
  MAX_TOOL_TURNS,
} from "../src/tools.js";

// ═══════════════════════════════════════════════════════════════
// Raw functions (the data sources)
// ═══════════════════════════════════════════════════════════════

describe("raw functions", () => {
  it("buildStatus returns data for a known service", () => {
    const result = JSON.parse(buildStatus("auth-service"));
    expect(result.status).toBe("healthy");
    expect(result.last_deploy).toBeTruthy();
  });

  it("buildStatus returns unknown for an unrecognized service", () => {
    const result = JSON.parse(buildStatus("nonexistent"));
    expect(result.status).toBe("unknown");
    expect(result.error).toBeTruthy();
  });

  it("recentDeploys respects the limit", () => {
    const result = JSON.parse(recentDeploys("payment-api", 2));
    expect(result.length).toBe(2);
  });

  it("activeIncidents returns an array", () => {
    expect(Array.isArray(JSON.parse(activeIncidents("auth-service")))).toBe(true);
  });
});

// ═══════════════════════════════════════════════════════════════
// Tool definitions
// ═══════════════════════════════════════════════════════════════

describe("getBuildStatus", () => {
  it("returns status for a known service", async () => {
    const result = JSON.parse(
      await getBuildStatus.invoke({ service_name: "auth-service" })
    );
    expect(result.status).toBe("healthy");
    expect(result.last_deploy).toBeTruthy();
  });

  it("returns unknown for an unrecognized service", async () => {
    const result = JSON.parse(
      await getBuildStatus.invoke({ service_name: "nonexistent" })
    );
    expect(result.status).toBe("unknown");
    expect(result.error).toBeTruthy();
  });
});

describe("getRecentDeploys", () => {
  it("returns a list of deployment records", async () => {
    const result = JSON.parse(
      await getRecentDeploys.invoke({ service_name: "payment-api", limit: 3 })
    );
    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBe(3);
    expect(result[0].status).toBe("success");
  });

  it("returns empty list for service with no deploys", async () => {
    const result = JSON.parse(
      await getRecentDeploys.invoke({ service_name: "inventory-service" })
    );
    expect(result).toEqual([]);
  });
});

describe("getActiveIncidents", () => {
  it("returns active incidents for payment-api", async () => {
    const result = JSON.parse(
      await getActiveIncidents.invoke({ service_name: "payment-api" })
    );
    expect(result.length).toBe(1);
    expect(result[0].id).toBe("INC-842");
  });

  it("returns empty list for service with no incidents", async () => {
    const result = JSON.parse(
      await getActiveIncidents.invoke({ service_name: "auth-service" })
    );
    expect(result).toEqual([]);
  });
});

// ═══════════════════════════════════════════════════════════════
// Failure injection
// ═══════════════════════════════════════════════════════════════

describe("flaky", () => {
  it("fails the first N calls then succeeds", () => {
    const fn = flaky(buildStatus, 2);
    expect(() => fn("auth-service")).toThrow(ConnectionError);
    expect(() => fn("auth-service")).toThrow(ConnectionError);
    expect(JSON.parse(fn("auth-service")).status).toBe("healthy");
  });

  it("isolates counters between wrappers", () => {
    const a = flaky(buildStatus, 1);
    const b = flaky(buildStatus, 1);
    expect(() => a("auth-service")).toThrow(ConnectionError);
    // b's counter is untouched by a's failure
    expect(() => b("auth-service")).toThrow(ConnectionError);
    expect(JSON.parse(a("auth-service")).status).toBe("healthy");
  });
});

// ═══════════════════════════════════════════════════════════════
// Tool execution
// ═══════════════════════════════════════════════════════════════

describe("executeToolSafely", () => {
  it("executes a valid tool call successfully", async () => {
    const result = JSON.parse(
      await executeToolSafely({
        name: "get_build_status",
        args: { service_name: "auth-service" },
      })
    );
    expect(result.status).toBe("healthy");
  });

  it("returns structured error for unknown tool", async () => {
    const result = JSON.parse(
      await executeToolSafely({ name: "nonexistent_tool", args: {} })
    );
    expect(result.error).toBeTruthy();
    expect(result.available_tools).toBeTruthy();
  });

  it("denies an unregistered tool via the registry whitelist", async () => {
    const registry = { get_build_status: {} }; // not even a real tool
    const result = JSON.parse(
      await executeToolSafely(
        { name: "delete_production_db", args: {} },
        0,
        { toolsByName: registry }
      )
    );
    expect(result.error).toBe("Unknown tool: 'delete_production_db'");
    expect(result.available_tools).toEqual(["get_build_status"]);
  });

  it("returns a structured error after retries are exhausted", async () => {
    const impl = flaky(buildStatus, 99);
    const flakyTool = tool(async ({ service_name }) => impl(service_name), {
      name: "get_build_status",
      description: "Return build status.",
      schema: z.object({ service_name: z.string() }),
    });

    const result = JSON.parse(
      await executeToolSafely(
        { name: "get_build_status", args: { service_name: "payment-api" } },
        1,
        { toolsByName: { get_build_status: flakyTool }, retryDelayMs: 0 }
      )
    );
    expect(result.status).toBe("failed");
    expect(result.attempts).toBe(2); // 1 initial + 1 retry
    expect(result.error).toBeTruthy();
  });
});

describe("ALL_TOOLS", () => {
  it("all tools have descriptions", () => {
    for (const t of ALL_TOOLS) {
      expect(t.description).toBeTruthy();
    }
  });

  it("TOOLS_BY_NAME maps all tools", () => {
    const names = ALL_TOOLS.map((t) => t.name).sort();
    expect(Object.keys(TOOLS_BY_NAME).sort()).toEqual(names);
  });

  it("exposes a bounded tool-turn limit", () => {
    expect(MAX_TOOL_TURNS).toBeGreaterThan(0);
  });
});

// ═══════════════════════════════════════════════════════════════
// Tool-calling loop (requires LLM)
// ═══════════════════════════════════════════════════════════════

describe("runToolLoop", () => {
  it("calls tool for build status question", async () => {
    const result = await runToolLoop("Is the auth-service healthy?", 0.0);
    expect(result.length).toBeGreaterThan(10);
    const lower = result.toLowerCase();
    expect(lower.includes("healthy") || lower.includes("auth")).toBe(true);
  }, 120_000);

  it("answers directly when no tool needed", async () => {
    const result = await runToolLoop("What is 2 + 2?", 0.0);
    expect(result).toContain("4");
  }, 120_000);
});

describe("runToolLoopWithTrace", () => {
  it("returns a trace with usage", async () => {
    const trace = await runToolLoopWithTrace("Is the payment-api healthy?", 0.0);
    expect(trace.answer).toBeTruthy();
    expect(trace.query).toBeTruthy();
    expect(trace.steps.length).toBeGreaterThanOrEqual(1);
    expect(trace.usage).toBeTruthy();
  }, 120_000);

  it("works when no tools are called", async () => {
    const trace = await runToolLoopWithTrace("Hello!", 0.0);
    expect(trace.answer).toBeTruthy();
  }, 120_000);
});
